Main.hs

```diff
@@ -2,28 +2,130 @@ module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
+import Language.Haskell.TH (Exp, safe)
+import Data.Either (isLeft, fromRight)
+import Control.Arrow (ArrowChoice(right))
 
-data Expr
+data Expr = Numb Double
+      | SqrtExpr Expr
+      | BinOpAdd Expr Expr
+      | BinOpSub Expr Expr
+      | BinOpMul Expr Expr
+      | BinOpDiv Expr Expr
+      | BinOpPow Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Numb a) = printf "%s" (show a)
+  show (SqrtExpr a) = printf "sqrt %s" (show a)
+  show (BinOpAdd a b) = printf "(%s + %s)" (show a) (show b)
+  show (BinOpSub a b) = printf "(%s - %s)" (show a) (show b)
+  show (BinOpMul a b) = printf "(%s * %s)" (show a) (show b)
+  show (BinOpDiv a b) = printf "(%s / %s)" (show a) (show b)
+  show (BinOpPow a b) = printf "(%s ^ %s)" (show a) (show b)
 
 instance Eq Expr where
-  (==) = undefined
+  Numb a == Numb b = a == b
+  SqrtExpr a == SqrtExpr b = a == b
+  BinOpAdd a b == BinOpAdd c d = (a == c) && (b == d)
+  BinOpSub a b == BinOpSub c d = (a == c) && (b == d)
+  BinOpMul a b == BinOpMul c d = (a == c) && (b == d)
+  BinOpDiv a b == BinOpDiv c d = (a == c) && (b == d)
+  BinOpPow a b == BinOpPow c d = (a == c) && (b == d)
+  _ == _ = False
 
-data Error
+data Error = DivByZero String | SqrtNeg String
```

Comment:
Rather than wrapping `String`s in the errors, consider using the expressions. This way, the code that creates the error doesn't have to deal with string conversion.